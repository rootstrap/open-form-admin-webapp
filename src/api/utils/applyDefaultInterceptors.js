import { saveSession, logout } from 'actions/userActions';

const ACCESS_TOKEN = 'access-token';
const UID = 'uid';
const CLIENT = 'client';
const RS_API_KEY = 'RS-API-KEY';

const UNAUTHORIZED = 401;

const defaultRequestInterceptors = store => [
  async request => {
    try {
      const { info } = store.getState().session;
      if (info) {
        const { token, client, uid } = info;
        request.headers = {
          ...request.headers,
          [ACCESS_TOKEN]: token,
          [RS_API_KEY]: process.env.RS_API_KEY,
          client,
          uid
        };
      }
    } catch (error) {
      console.log('Failed to load session headers', error); // eslint-disable-line
    }
    return request;
  }
];

const defaultResponseInterceptors = store => [
  async response => {
    if (response.ok) {
      const { headers } = response;
      const token = headers.get(ACCESS_TOKEN);
      if (token) {
        const session = {
          token,
          uid: headers.get(UID),
          client: headers.get(CLIENT)
        };
        store.dispatch(saveSession(session));
      }
    }
    if (response.status === UNAUTHORIZED) {
      store.dispatch(logout());
    }
    return response;
  }
];

export default (store, apiService) => {
  defaultRequestInterceptors(store).forEach(interceptor =>
    apiService.requestInterceptors.use(interceptor)
  );

  defaultResponseInterceptors(store).forEach(interceptor =>
    apiService.responseInterceptors.use(interceptor)
  );
};
