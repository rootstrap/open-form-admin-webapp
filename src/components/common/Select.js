import React from 'react';
import styled, { css } from 'styled-components';
import PropTypes from 'prop-types';

import DropdownArrow from 'components/common/DropdownArrow';

const StyledSelect = styled.select`
  flex: 1;
  border-width: 0;
  margin: 0.1em 0;
  padding: 0.5em 0.7em;
  appearance: none;
  height: 40px;
  font-size: 0.9em;
  ${({ theme }) => css`
    background-color: ${theme.background.input};
    border-radius: ${theme.border.radius};
  `}
`;

const Wrapper = styled.div`
  display: flex;
  align-items: center;
  position: relative;
`;

const StyledDropdownArrow = styled(DropdownArrow)`
  position: absolute;
  right: 0.7em;
`;

const Select = ({ children, ...props }) => (
  <Wrapper>
    <StyledSelect {...props}>{children}</StyledSelect>
    <StyledDropdownArrow />
  </Wrapper>
);

Select.propTypes = {
  children: PropTypes.node.isRequired
};

export default Select;
