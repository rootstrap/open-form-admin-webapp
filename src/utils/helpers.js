import queryString from 'query-string';

export const parseInputErrors = error => {
  if (!error) {
    return;
  }
  if (Array.isArray(error)) {
    return error[0];
  }
  return error;
};

export const applyQueryParams = (url, params = {}) => {
  if (!Object.keys(params).length) {
    return url;
  }
  const queryParams = queryString.stringify(params);
  return `${url}?${queryParams}`;
};
