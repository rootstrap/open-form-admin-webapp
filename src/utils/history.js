import createBrowserHistory from 'history/createBrowserHistory';
import createMemoryHistory from 'history/createMemoryHistory';

let instance;

class History {
  constructor() {
    if (!instance) {
      instance = process.env.BROWSER ? createBrowserHistory() : createMemoryHistory();
    }
    return instance;
  }
}

export default new History();
