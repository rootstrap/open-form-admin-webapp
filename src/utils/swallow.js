const swallow = (fail, fn) => (...args) => {
  try {
    return fn(...args);
  } catch (error) {
    return fail(error, ...args);
  }
};

export default swallow;
