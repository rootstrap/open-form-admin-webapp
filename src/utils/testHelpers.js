import React from 'react';
import TestRenderer from 'react-test-renderer';
import { ThemeProvider } from 'styled-components';

import theme from 'constants/theme';

export function renderWithTheme(component) {
  return TestRenderer.create(<ThemeProvider theme={theme}>{component}</ThemeProvider>);
}
