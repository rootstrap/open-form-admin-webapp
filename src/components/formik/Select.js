import React from 'react';
import { useField } from 'formik';
import PropTypes from 'prop-types';

import Select from 'components/common/Select';
import FormGroup from 'components/common/FormGroup';
import Label from 'components/common/Label';

const FormikSelect = ({ name, label, ...props }) => {
  const [field] = useField(name);
  return (
    <FormGroup>
      <Label htmlFor={name}>{label}</Label>
      <Select
        {...props}
        {...field}
        onChange={({ value }) => field.onChange({ target: { value, name } })}
      />
    </FormGroup>
  );
};

FormikSelect.propTypes = {
  name: PropTypes.string.isRequired,
  label: PropTypes.oneOfType([PropTypes.string, PropTypes.node]).isRequired
};

export default FormikSelect;
