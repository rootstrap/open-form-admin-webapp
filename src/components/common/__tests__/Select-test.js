import React from 'react';

import { renderWithTheme } from 'utils/testHelpers';
import Select from '../Select';

it('renders correctly', () => {
  const testRenderer = renderWithTheme(
    <Select
      options={[{ label: 'option1', value: '1' }, { label: 'option2', value: '2' }]}
      value="1"
      onChange={() => {}}
    />
  );
  expect(testRenderer.toJSON()).toMatchSnapshot();
});

it('changes option', () => {
  const mockCallback = jest.fn();
  const testRenderer = renderWithTheme(
    <Select
      options={[{ label: 'option1', value: '1' }, { label: 'option2', value: '2' }]}
      value="1"
      onChange={mockCallback}
    />
  );
  const testInstance = testRenderer.root;
  testInstance.findByProps({ children: 'option2' }).props.onClick();
  expect(mockCallback.mock.calls[0][0].value).toBe('2');
});
