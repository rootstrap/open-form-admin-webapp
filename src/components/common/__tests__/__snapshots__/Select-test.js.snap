// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
.c2.enter {
  -webkit-transform: rotate(0deg);
  -ms-transform: rotate(0deg);
  transform: rotate(0deg);
}

.c2.enter-active {
  -webkit-transform: rotate(180deg);
  -ms-transform: rotate(180deg);
  transform: rotate(180deg);
  -webkit-transition: -webkit-transform 250ms;
  -webkit-transition: transform 250ms;
  transition: transform 250ms;
}

.c2.enter-done {
  -webkit-transform: rotate(180deg);
  -ms-transform: rotate(180deg);
  transform: rotate(180deg);
}

.c2.exit {
  -webkit-transform: rotate(180deg);
  -ms-transform: rotate(180deg);
  transform: rotate(180deg);
}

.c2.exit-active {
  -webkit-transform: rotate(0deg);
  -ms-transform: rotate(0deg);
  transform: rotate(0deg);
  -webkit-transition: -webkit-transform 250ms;
  -webkit-transition: transform 250ms;
  transition: transform 250ms;
}

.c3 {
  position: absolute;
  right: 0.7rem;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
  height: 40px;
}

.c1 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  border-width: 0;
  margin: 0.1em 0;
  padding: 0.5rem 0.7rem;
  height: 40px;
  font-size: 0.9em;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: #f2f2f2;
  border-radius: 2px;
}

<div
  className="c0"
  data-testid="toggle"
  onClick={[Function]}
>
  <div
    className="c1"
  >
    option1
  </div>
  <svg
    className="c2 c3"
    fill="none"
    height="8"
    viewBox="0 0 15 8"
    width="15"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M1.48242 1L7.48242 7L13.4824 1"
      stroke="#4F4F4F"
      strokeLinecap="round"
      strokeLinejoin="round"
      strokeWidth="2"
    />
  </svg>
</div>
`;
